### PR TITLE
SMC 182 , SMC 114 - Convert criteria to elasticsearch criteria

### PR DESCRIPTION
**Pull request by ErasmusJW**
SMC 114 : Convert criteria to elasticsearch criteria. Sift module no longer used to filter data after elastic search return all data that matches path. 

Github issue : https://github.com/happner/elastic-dataprovider/issues/14
Jira Issue : http://jira.msasafety.com/browse/SMC-114

SMC-182 : Make count compatible with happner
Github issue https://github.com/happner/elastic-dataprovider/issues/13

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -23,21 +20,20 @@ function ElasticProvider(config) {
...
   Object.defineProperty(this, '__comedian', {value: new Comedian(config.wildcardCache)});
-
 }
 
 ElasticProvider.prototype.UPSERT_TYPE = {
   upsert: 0,
   update: 1,
   insert: 2,
-  bulk: 3
+  bulk: 3,
```
Comment (cordellcalitz) [inline index.js]: lint error?

File: lib/mongo-to-elastic.js [hunk 2]
```diff
@@ -1,5 +1,252 @@
...
+      if(value.length > 0) {
+        let valueString = "(" + value[0]
+        for (let val = 1; val < value.length; ++val )
+        {
+          valueString += " OR " + value[val];
+        }
+        valueString += ")"
+        return `${field}:${valueString}`
+      } else
+        throw new Error(`Comparison Query malformed: Expected a Populated Array value for ${key}`);
```
Comment (cordellcalitz) [inline lib/mongo-to-elastic.js]: Rather swap around check and exit early than have an else

File: index.js [hunk 3]
```diff
@@ -384,279 +360,293 @@ function __createBulkMessage(path, setData, options) {
...
     elasticMessage.body = {
-      "query": {
-        "wildcard": {
-          "path": path
-        }
-      }
+      'query': {
+        'wildcard': {
+          'path': path,
+        },
```
Comment (cordellcalitz) [inline index.js]: lots of unnecessary commas added everywhere

File: test/func.js [hunk 4]
```diff
@@ -40,6 +44,35 @@ describe('func', function () {
 
   var serviceInstance = new service(config);
 
+  // note that  modifies the path
```
Comment (cordellcalitz) [inline test/func.js]: what modifies the path?

File: test/func.js [hunk 5]
```diff
@@ -401,16 +410,40 @@ describe('func', function () {
...
             }
 
-            done();
+            //ascending
+            randomItems.sort(function (a, b) {
+
+              return b.item_sort_id - a.item_sort_id;
+            });
+
+            serviceInstance.find(base_path + '/*', {
```
Comment (cordellcalitz) [inline test/func.js]: Why does this test have /* but the ascending test only had *

File: test/func.js [hunk 6]
```diff
@@ -521,4 +554,585 @@ describe('func', function () {
...
+      {data: {"test": "data2"}},
+      {data: {"test": "data2"}}
+    ]
+
+    let insertCount = 0
+
+    async function findEntries()
+    {
+
+      // serviceInstance.findNew("/findTest/num1",{criteria : {"query":"data.test:data1"}},(err,data)=>{
```
Comment (cordellcalitz) [inline test/func.js]: dead code

File: test/func.js [hunk 7]
```diff
@@ -521,4 +554,585 @@ describe('func', function () {
...
+      })
+
+    }
+
+
+    for(let i = 0 ; i < dataItem.length ; ++i)
+    {
+      serviceInstance.upsert(`/findTest/num${i}`,dataItem[i],{},false,(err)=>{
+        if(err)
+          done(err)
```
Comment (cordellcalitz) [inline test/func.js]: return to avoid done called more than once

File: test/func.js [hunk 8]
```diff
@@ -521,4 +554,585 @@ describe('func', function () {
...
+                "query":     [{'path' :'/findTest/num1' }],
+              }
+            }
+          }
+        }
+      };
+
+      serviceInstance.find("/findTest/num1",body,(err,data)=>{
+
+        done();
```
Comment (cordellcalitz) [inline test/func.js:694]: Isn't there a test condition missing? Some expect for no error and some data?

File: test/func.js [hunk 9]
```diff
@@ -521,4 +554,585 @@ describe('func', function () {
...
+
+    AddSearchDelete("/criteriaConversionsNe",dataItemAdded,filterItemCorrect,filterItemIncorect).then((data)=>{
+
+      return  AddSearchDelete("/criteriaConversionsNe",dataItemAdded,filterItemCorrect2,filterItemIncorect2)
+    }).then((data)=>{
+      done();
+    }).catch(done)
+  })
+
+  it('Criteria Conversion - Mongo Operator in  ',  function(done) {
```
Comment (cordellcalitz) [inline test/func.js]: Is this a duplicate test? There is another in test above

File: test/func.js [hunk 10]
```diff
@@ -521,4 +554,585 @@ describe('func', function () {
...
+              [ [ 0, 0 ], [ 3, 6 ], [ 6, 1 ], [ 0, 0 ] ]
+            ]
+          }
+        }
+      }
+    }
+
+
+    AddSearchDelete("/criteriaConversionsOr",dataItemAdded,invalidfilter,{}).then((data)=>{
+      done(data);
```
Comment (cordellcalitz) [inline test/func.js]: unexpected behavior - calling done before returning another promise

**Review by cordellcalitz — APPROVED** [review]

File: test/func.js (hunk 4 above)
Comment (ErasmusJW) [reply test/func.js]: more verbose comment added

File: test/func.js (hunk 5 above)
Comment (ErasmusJW) [reply test/func.js]: fixed

File: test/func.js (hunk 6 above)
Comment (ErasmusJW) [reply test/func.js]: fixed

File: test/func.js (hunk 7 above)
Comment (ErasmusJW) [reply test/func.js]: fixed

File: test/func.js (hunk 8 above)
Comment (ErasmusJW) [reply test/func.js:694]: added missing checks

File: test/func.js (hunk 9 above)
Comment (ErasmusJW) [reply test/func.js]: duplicate removes

File: test/func.js (hunk 10 above)
Comment (ErasmusJW) [reply test/func.js]: fixed

**Review by cordellcalitz — APPROVED** [review]

File: lib/mongo-to-elastic.js [hunk 11]
```diff
@@ -1,20 +1,233 @@
...
+      for (let val of value) {
+        valueString += ' AND NOT' + val;
+      }
+      valueString += ' OR NOT _exists_)';
+      return `${field}:${valueString}`;
+    }
+  ]
+]);
+
+const monogLogicalOperators = new Map([
```
Comment (southbite) [inline lib/mongo-to-elastic.js:45]: typo

**Review by csampsonza — CHANGES_REQUESTED** [review]
Please add a coverage report

**Comment by cordellcalitz** [comment]
> Please add a coverage report

Remember this is not an SMC repo, I'm not sure what the requirements are here

File: lib/mongo-to-elastic.js [hunk 12]
```diff
@@ -1,20 +1,233 @@
...
+  let returnString = '';
+  switch (typeof value) {
+    case 'string':
+      return `${key}:${formatStringCondition(value)}`;
+
+    case 'object':
+      if (value.constructor.name === 'RegExp') return `${key}:${value}`;
+      for (const [subKey, subValueValue] of Object.entries(value)) {
+        if (returnString !== '') returnString += ' AND ';
+        if (isMonogOpperator(subKey)) {
```
Comment (southbite) [inline lib/mongo-to-elastic.js]: typo

File: lib/mongo-to-elastic.js [hunk 13]
```diff
@@ -1,20 +1,233 @@
...
+    field = keyParrent + '.';
+  }
+  field += key;
+  return ProcessValueField(field, value);
+}
+
+function AddCriteria(criteriaKey, criteriaValue, keyParrent = '', field = '') {
+  let queryString = '( ';
+
+  if (!isMonogOpperator(criteriaKey)) {
```
Comment (southbite) [inline lib/mongo-to-elastic.js]: typo

File: lib/mongo-to-elastic.js [hunk 14]
```diff
@@ -1,20 +1,233 @@
...
+    (field, value) => {
+      if (field === '')
+        throw new Error("Evaluation Query malfomed: '$regex' requires a field value");
+      if (value.constructor.name === 'RegExp') return `${field}:${value}`;
+      else return `${field}:${formatStringCondition(value)}`;
+    }
+  ]
+]);
+
+function isMonogOpperator(string) {
```
Comment (southbite) [inline lib/mongo-to-elastic.js]: typo

File: lib/mongo-to-elastic.js [hunk 15]
```diff
@@ -1,20 +1,233 @@
...
+  if (keyParrent !== '') fieldComplete = keyParrent + '.';
+
+  fieldComplete += field;
+  let comparisonOperation = mongoComparisonOperators.get(key);
+  if (comparisonOperation !== undefined) {
+    if (fieldComplete === '')
+      throw new Error(`Comparison Query malformed: Expected a field  value for ${key}`);
+    else return comparisonOperation(fieldComplete, value);
+  }
+  let logicalOpperation = monogLogicalOperators.get(key);
```
Comment (southbite) [inline lib/mongo-to-elastic.js:143]: typo

**Review by southbite — APPROVED** [review]

**Pull request merged**